pkg/util/archive/zip.go

```diff
@@ -0,0 +1,261 @@
+	destAbs, err := filepath.Abs(destination)
+	if err != nil {
+		return fmt.Errorf("%w getting absolute path of destination %s: %s", err, destination, source)
+	}
+
+	return filepath.Walk(source, func(fpath string, info os.FileInfo, err error) error {
+		if err != nil {
+			return fmt.Errorf("%w traversing %s", err, fpath)
+		}
+		if info == nil {
+			return fmt.Errorf("%s: no file info", fpath)
+		}
+
+		fpathAbs, err := filepath.Abs(fpath)
+		if err != nil {
+			return fmt.Errorf("%w %s: getting absolute path", err, fpath)
+		}
+		if within(fpathAbs, destAbs) {
+			return nil
+		}
+
+		// build the name to be used within the archive
+		nameInArchive, err := makeNameInArchive(sourceInfo, source, "", fpath)
+		if err != nil {
+			return err
+		}
+
+		if info.Mode().IsRegular() {
+			file, err := os.Open(fpath)
+			if err != nil {
+				return fmt.Errorf("%w %s: opening", err, fpath)
+			}
+			defer file.Close()
+
+			finfo := fileInfo{
+				FileInfo:   info,
+				customName: nameInArchive,
+			}
+			header, err := zip.FileInfoHeader(finfo)
+			if err != nil {
+				return fmt.Errorf("%w %s: getting header", err, finfo.Name())
+			}
+
+			if finfo.IsDir() {
+				header.Name += "/"
+			}
+			header.Method = zip.Store
+
+			writer, err := zipW.CreateHeader(header)
+			if err != nil {
+				return fmt.Errorf("%w %s: making header", err, finfo.Name())
+			}
+
+			if finfo.IsDir() {
+				return nil
+			}
+
+			_, err = io.Copy(writer, file)
+			if err != nil {
+				return fmt.Errorf("%w %s: copying contents", err, finfo.Name())
+			}
+		}
+
+		return nil
+	})
+}
+
+// makeNameInArchive is an adapted implementation of makeNameInArchive from
+// https://github.com/mholt/archiver/blob/v3.5.1/archiver.go#L413
+// Copyright (c) 2016 Matthew Holt
+//
+// makeNameInArchive returns the filename for the file given by fpath to be used within
+// the archive. sourceInfo is the FileInfo obtained by calling os.Stat on source, and baseDir
+// is an optional base directory that becomes the root of the archive. fpath should be the
+// unaltered file path of the file given to a filepath.WalkFunc.
+func makeNameInArchive(sourceInfo os.FileInfo, source, baseDir, fpath string) (string, error) {
+	name := filepath.Base(fpath) // start with the file or dir name
+	if sourceInfo.IsDir() {
+		// preserve internal directory structure; that's the path components
+		// between the source directory's leaf and this file's leaf
+		dir, err := filepath.Rel(filepath.Dir(source), filepath.Dir(fpath))
+		if err != nil {
+			return "", err
+		}
+		// prepend the internal directory structure to the leaf name,
+		// and convert path separators to forward slashes as per spec
+		name = path.Join(filepath.ToSlash(dir), name)
+	}
+	return path.Join(baseDir, name), nil // prepend the base directory
+}
+
+// This function is inspired by https://github.com/mholt/archiver/blob/v3.5.1/archiver.go#L360
+// within returns true if sub is within or equal to parent.
+func within(parent, sub string) bool {
```

Comment:
Maybe something that's also doable via [securejoin](https://github.com/cyphar/filepath-securejoin)?

Reply:
Non-blocking nit if we keep this:
```suggestion
func isPathWithinParentPath(parentPath string, childPath string) bool {
```
or something similar